rust/actyx/ax-aql/src/language/type_check.rs

```diff
@@ -184,55 +183,365 @@ impl<'a> QueryTypeCheck<'a> {
 /// Check whether Ind can be used to index a Type
 /// e.g. can _.a.b be used to access Type::Record([("a", Type::Record([("b", Type::String)]))])
 /// Drills down type and Ind.tail recursively at the same time to find a match
-fn index_tail_access_matches(cur_type: &Type, Ind { tail, .. }: &Ind) -> Result<(), ExtraMessage> {
-    fn recurse(cur_type: &Type, index: &[Index]) -> Result<(), ExtraMessage> {
-        let (first, rest) = index.split_at(1);
-        if first.is_empty() {
-            return Ok(());
+fn drill_down_type_by_index(cur_type: &Type, Ind { head: _, tail, .. }: &Ind) -> Result<Type, ExtraMessage> {
+    fn recurse(cur_type: &Type, index: &[Index]) -> Result<Type, ExtraMessage> {
+        if index.len() < 1 {
+            return Ok(cur_type.to_owned());
         }
-
+        let (first, rest) = index.split_at(1);
         let first = &first[0];
 
         match cur_type {
-            Type::Atom(x) => match x {
-                crate::TypeAtom::Universal => Ok(()),
-                crate::TypeAtom::Tuple(x) => match first {
-                    Index::Number(num_index) => {
-                        if (*num_index as usize) < x.len() {
-                            Ok(())
-                        } else {
-                            Err(format!("{:?} cannot be accessed by {}", x, num_index))
-                        }
-                    }
-                    _ => Err(format!("{:?} cannot be accessed by {:?}", x, first)),
+            Type::Atom(type_atom) => match type_atom {
```

Comment:
Nitpick: I'd make this `Type::Atom(TypeAtom::Universal)` (for example) so you could avoid a whole level of nesting and get "extra context" when reading the match arm